emma-language/src/main/scala/org/emmalanguage/compiler/lang/opt/FoldGroupFusion.scala

```diff
@@ -0,0 +1,249 @@
+package compiler.lang.opt
+
+import compiler.Common
+import compiler.lang.cf.ControlFlow
+import compiler.lang.core.Core
+import util.Graphs.topoSort
+
+import quiver.LNode
+
+import scala.collection.SortedMap
+import scala.collection.breakOut
+
+/** The fold-fusion optimization. */
+private[compiler] trait FoldGroupFusion extends Common {
+  this: Core with ControlFlow =>
+
+  /** The fold-fusion optimization. */
+  object FoldGroupFusion {
+
+    private type Node = LNode[u.TermSymbol, u.ValDef]
+    private type MMap = Map[u.TermSymbol, Match]
+
+    import API._
+    import Core.{Lang => core}
+    import UniverseImplicits._
+
+    val cs = new Comprehension.Syntax(DataBag.sym)
+
+    /**
+     * Performs the fold-group-fusion optimization on `DataBag` expressions.
+     *
+     * Looks for patterns of the following shape:
+     *
+     * {{{
+     *   //...
+     *   val x = y.groupBy(k)
+     *   //...
+     *   val c = for (g <- x) yield {
+     *     //...
+     *     // <dependencies of `a` inside `c`>
+     *     val v = g.values
+     *     val f = v.fold(a)
+     *     //...
+     *   }
+     * }}}
+     *
+     * where
+     *
+     * - the only use of `x` is in `c`;
+     * - the only use of `v` is in `f`;
+     * - the the algebra `a` does not (transitively) depend on `g`.
+     *
+     * Matches are rewritten as follows.
+     *
+     * {{{
+     *   //...
+     *   // <dependencies of `a` inside `c`>
+     *   val x = LocalOps.foldGroup(y, k, a)
+     *   //...
+     *   val c = for (g <- x) yield {
+     *     // ...
+     *     val f = g.values
+     *     // ...
+     *   }
+     * }}}
+     */
+    lazy val foldGroupFusion: u.Tree => u.Tree = tree => {
+      val cfg = CFG.graph(tree)
+
+      val ms = for {
+        GroupBy(x, y, k) <- cfg.data.labNodes.map(_.label)
+        if cfg.uses(x) == 1
+        xUses = cfg.data.outEdges(x).flatMap(e => cfg.data.label(e.to))
+        GroupComprehension(c, g, `x`) <- xUses
+        gUses = cfg.data.outEdges(g).flatMap(e => cfg.data.label(e.to))
+        ProjValues(v, g) <- gUses
+        if cfg.uses(v) == 1
+        vUses = cfg.data.outEdges(v).flatMap(e => cfg.data.label(e.to))
+        FoldValues(f, v, a) <- vUses
+        deps = cfg.data.reverse
+        d = deps.reachable(a).toSet
+        if !(d contains g)
+      } yield {
+        val C = cfg.nest.outs(c).map(_._2).toSet
+        val dm = SortedMap((for {
+          v <- deps.nodes
+          if d contains v
+        } yield v -> deps.outs(v).map(_._2).toSet): _*)(ordTermSymbol)
+        val ds = for {
+          v <- topoSort(dm).get
+          if C contains v
+          w <- cfg.data.label(v)
+        } yield w
+        Match(x, y, k, c, g, v, f, a, ds)
+      }
+
+      val xMap: MMap = ms.map(c => c.x -> c)(breakOut)
+      val gMap: MMap = ms.map(c => c.g -> c)(breakOut)
+      val vMap: MMap = ms.map(c => c.v -> c)(breakOut)
+      val fMap: MMap = ms.map(c => c.f -> c)(breakOut)
+
+      api.TopDown.unsafe.transform({
```

Comment:
It's a good idea to add a comment why this is `unsafe`:
* `Group[K, DataBag[A]]` becomes `Group[K, B]`